test/vegeta/pacers/combined_pacer.go

```diff
@@ -0,0 +1,120 @@
+	durations []time.Duration
+
+	// totalDuration is sum of the given Durations.
+	totalDuration uint64
+	// stepDurations is the accumulative duration of each step calculated from the given Durations.
+	stepDurations []uint64
+
+	curtPacerIndex  uint
+	prevElapsedHits uint64
+	prevElapsedTime uint64
+}
+
+// NewCombined returns a new CombinedPacer with the given config.
+func NewCombined(pacers []vegeta.Pacer, durations []time.Duration) (vegeta.Pacer, error) {
+	if len(pacers) == 0 || len(durations) == 0 || len(pacers) != len(durations) || len(pacers) == 1 {
+		return nil, errors.New("configuration for this CombinedPacer is invalid")
+	}
+
+	var totalDuration uint64
+	stepDurations := make([]uint64, len(pacers))
+	for i, duration := range durations {
+		totalDuration += uint64(duration)
+		if i == 0 {
+			stepDurations[i] = uint64(duration)
+		} else {
+			stepDurations[i] = stepDurations[i-1] + uint64(duration)
+		}
+	}
+	pacer := &combinedPacer{
+		pacers:    pacers,
+		durations: durations,
+
+		totalDuration: totalDuration,
+		stepDurations: stepDurations,
+
+		curtPacerIndex:  0,
+		prevElapsedHits: 0,
+		prevElapsedTime: 0,
+	}
+	return pacer, nil
+}
+
+// combinedPacer satisfies the Pacer interface.
+var _ vegeta.Pacer = &combinedPacer{}
+
+// String returns a pretty-printed description of the combinedPacer's behaviour.
+func (cp *combinedPacer) String() string {
+	var sb strings.Builder
+	for i := range cp.pacers {
+		pacerStr := fmt.Sprintf("Pacer: %s, Duration: %s\n", cp.pacers[i], cp.durations[i])
+		sb.WriteString(pacerStr)
+	}
+	return sb.String()
+}
+
+func (cp *combinedPacer) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	pacerTimeOffset := uint64(elapsedTime) % cp.totalDuration
+	pacerIndex := cp.getPacerIndex(pacerTimeOffset)
+	// The pacer must be the same or the next neighbor of the current pacer, otherwise stop the attack.
+	if pacerIndex != cp.curtPacerIndex &&
+		pacerIndex != cp.curtPacerIndex+1 &&
+		!(pacerIndex == 0 && cp.curtPacerIndex == uint(len(cp.pacers)-1)) {
+		return 0, true
+	}
+
+	// If it needs to switch to the next pacer, update prevElapsedTime, prevElapsedHits and curtPacerIndex.
+	if pacerIndex != cp.curtPacerIndex {
+		cp.prevElapsedTime = uint64(elapsedTime)
+		cp.prevElapsedHits = elapsedHits
+		cp.curtPacerIndex = pacerIndex
+	}
+
+	// Use the adjusted elapsedTime and elapsedHits to get the time to wait for the next hit.
+	curtPacer := cp.pacers[cp.curtPacerIndex]
+	curtElapsedTime := time.Duration(uint64(elapsedTime) - cp.prevElapsedTime)
+	curtElapsedHits := elapsedHits - cp.prevElapsedHits
+	return curtPacer.Pace(curtElapsedTime, curtElapsedHits)
+}
+
+// getPacerIndex returns the index of pacer that pacerTimeOffset falls into
+func (cp *combinedPacer) getPacerIndex(pacerTimeOffset uint64) uint {
+	for i, stepDuration := range cp.stepDurations {
+		if pacerTimeOffset < stepDuration {
```

Comment:
This can be binary search, if we care to be fast...

Reply:
Changed to binary search. Please don't fail my interview :-)